ovirt/resource_ovirt_affinity_group.go

```diff
@@ -0,0 +1,425 @@
+		vmRuleUpdate = true
+	}
+	if vmRuleUpdate {
+		group.VmsRule(vmRuleBuilder.MustBuild())
+		attributeUpdate = true
+	}
+
+	hostRuleBuilder := ovirtsdk4.NewAffinityRuleBuilder()
+	hostRuleUpdate := false
+	if d.HasChange("host_positive") {
+		hostRuleBuilder.Positive(d.Get("host_positive").(bool))
+		hostRuleUpdate = true
+	}
+	if d.HasChange("host_enforcing") {
+		hostRuleBuilder.Enforcing(d.Get("host_enforcing").(bool))
+		hostRuleUpdate = true
+	}
+	if d.HasChange("host_list") {
+		hostRuleBuilder.Enabled(len(d.Get("host_list").([]interface{})) > 0)
+		hostRuleUpdate = true
+	}
+	if hostRuleUpdate {
+		group.HostsRule(hostRuleBuilder.MustBuild())
+		attributeUpdate = true
+	}
+
+	if attributeUpdate {
+		log.Printf("[DEBUG] Updating %#v", group.MustBuild())
+		_, err := conn.SystemService().
+			ClustersService().
+			ClusterService(d.Get("cluster_id").(string)).
+			AffinityGroupsService().
+			GroupService(d.Id()).
+			Update().
+			Group(group.MustBuild()).
+			Send()
+		if err != nil {
+			log.Printf("[DEBUG] Error updating affinity group (%s): %s", d.Id(), err)
+			return err
+		}
+	}
+
+	if d.HasChange("vm_list") {
+		vmsService := conn.SystemService().
+			ClustersService().
+			ClusterService(d.Get("cluster_id").(string)).
+			AffinityGroupsService().
+			GroupService(d.Id()).
+			VmsService()
+
+		if err := updateVmList(vmsService, d.Get("vm_list").([]interface{})); err != nil {
+			return err
+		}
+	}
+
+	if d.HasChange("host_list") {
+		hostsService := conn.SystemService().
+			ClustersService().
+			ClusterService(d.Get("cluster_id").(string)).
+			AffinityGroupsService().
+			GroupService(d.Id()).
+			HostsService()
+
+		if err := updateHostList(hostsService, d.Get("host_list").([]interface{})); err != nil {
+			return err
+		}
+	}
+
+	return resourceOvirtAffinityGroupRead(d, meta)
+}
+
+func resourceOvirtAffinityGroupDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*ovirtsdk4.Connection)
+	_, err := conn.SystemService().
+		ClustersService().
+		ClusterService(d.Get("cluster_id").(string)).
+		AffinityGroupsService().
+		GroupService(d.Id()).
+		Remove().
+		Send()
+	if err != nil {
+		if _, ok := err.(*ovirtsdk4.NotFoundError); ok {
+			return nil
+		}
+		return err
+	}
+	return nil
+}
+
+func stringInSlice(a string, list []interface{}) bool {
+	for _, b := range list {
+		if b.(string) == a {
+			return true
+		}
+	}
+	return false
+}
+
+func updateVmList(vmsService *ovirtsdk4.AffinityGroupVmsService, vmList []interface{}) error {
+	// Add VMs to affinity group
+	if vms, ok := vmsService.List().MustSend().Vms(); ok {
+		currentVms := vms.Slice()
+
+		// Basically implement set subtraction on both sides
+		for _, v := range currentVms {
+			if !stringInSlice(v.MustId(), vmList) {
+				log.Printf("[DEBUG] Removing VM %s from affinity group", v.MustId())
+				if _, err := vmsService.VmService(v.MustId()).Remove().Send(); err != nil {
+					return err
+				}
+			}
+		}
+
+		for _, v := range vmList {
```

Comment:
Seems that if vmList (the new VMs list) have removed and also added Vms comparing to the currentVms list, you can end up with an error and return from the 2nd for block while you still succeed to remove items in teh 1st loop , resulting in inconsistent state since the operation should fully succeed or fully fail